nimbess-cni.go

```diff
@@ -0,0 +1,314 @@
+		"ContainerID": args.ContainerID,
+		"Netns":       args.Netns,
+		"IfName":      args.IfName,
+		"Args":        args.Args,
+	}).Debug("CNI ADD request")
+
+	// Prepare CNI Request for Network Config
+	cniRequestNW := &cninimbess.CNIRequest_NetworkConfig{}
+
+	// Prepare CNI request
+	cniRequest := &cninimbess.CNIRequest{
+		Version:          conf.CNIVersion,
+		ContainerId:      args.ContainerID,
+		NetworkNamespace: args.Netns,
+		InterfaceName:    args.IfName,
+		NetworkConfig:    cniRequestNW,
+		ExtraArguments:   cniRequestNW.Args,
+	}
+
+	cniResult := &cnitypes.Result{
+		CNIVersion: conf.CNIVersion,
+	}
+
+	// TODO: handle IPAM type and data
+
+	// Connect to the remote CNI handler over gRPC
+	conn, c, err := grpcConnect(conf.GrpcServer)
+	if err != nil {
+		log.Errorf("Unable to connect to GRPC server %s: %v", conf.GrpcServer, err)
+		return err
+	}
+	defer conn.Close()
+
+	// Execute the remote ADD request
+	r, err := c.Add(context.Background(), cniRequest)
+	if err != nil {
+		log.Errorf("Error by executing remote CNI Add request: %v", err)
+		return err
+	}
+
+	// Process interfaces
+	for ifidx, iface := range r.Interfaces {
+		// append interface info
+		cniResult.Interfaces = append(cniResult.Interfaces, &cnitypes.Interface{
+			Name:    iface.Name,
+			Mac:     iface.Mac,
+			Sandbox: iface.Sandbox,
+		})
+		for _, ip := range iface.IpAddresses {
+			// append interface ip address info
+			_, ipAddr, err := net.ParseCIDR(ip.Address)
+			if err != nil {
+				log.Error(err)
+				return err
+			}
+			var gwAddr net.IP
+			if ip.Gateway != "" {
+				gwAddr = net.ParseIP(ip.Gateway)
+				if err != nil {
+					log.Error(err)
+					return err
+				}
+			}
+			ver := "4"
+			if ip.Version == cninimbess.CNIReply_Interface_IP_IPV6 {
+				ver = "6"
+			}
+			cniResult.IPs = append(cniResult.IPs, &cnitypes.IPConfig{
+				Address:   *ipAddr,
+				Version:   ver,
+				Interface: &ifidx,
+				Gateway:   gwAddr,
+			})
+		}
+	}
+
+	// Process routes
+	for _, route := range r.Routes {
+		_, dstIP, err := net.ParseCIDR(route.Dst)
+		if err != nil {
+			log.Error(err)
+			return err
+		}
+		gwAddr := net.ParseIP(route.Gw)
+		if err != nil {
+			log.Error(err)
+			return err
+		}
+		cniResult.Routes = append(cniResult.Routes, &types.Route{
+			Dst: *dstIP,
+			GW:  gwAddr,
+		})
+	}
+
+	// Process DNS entry
+	for _, dns := range r.Dns {
+		cniResult.DNS.Nameservers = dns.Nameservers
+		cniResult.DNS.Domain = dns.Domain
+		cniResult.DNS.Search = dns.Search
+		cniResult.DNS.Options = dns.Options
+	}
+
+	log.WithFields(log.Fields{"Result": cniResult}).Debugf("CNI ADD request OK, took %s", time.Since(start))
+
+	return cniResult.Print()
+}
+
+// cmdDel implements the CNI request to delete a container from network.
+// It forwards the request to he remote gRPC server and returns the result received from gRPC.
+func cmdDel(args *skel.CmdArgs) error {
+	start := time.Now()
+
+	// Parse CNI config
+	conf, err := parseCNIConfig(args.StdinData)
```

Comment:
move some of this to a common function since it is duplicated?

Reply:
I've already created the parseCNIConfig bits in a function, that is called twice.

Reply:
sorry, I meant the first few lines of this function are the same as the previous function = start := time, parseCni, ... all look the same as the previous function so thought it could be moved to a common function.